library/std/src/panicking.rs

```diff
@@ -336,20 +336,26 @@ pub mod panic_count {
     pub fn increase() -> (bool, usize) {
         let global_count = GLOBAL_PANIC_COUNT.fetch_add(1, Ordering::Relaxed);
         let must_abort = global_count & ALWAYS_ABORT_FLAG != 0;
-        let panics = if must_abort {
-            global_count & !ALWAYS_ABORT_FLAG
-        } else {
-            LOCAL_PANIC_COUNT.with(|c| {
+        let mut panics = global_count & !ALWAYS_ABORT_FLAG;
+
+        // In a restricted_std environment, we likely don't have thread_local.
+        // If we attempt to use LOCAL_PANIC_COUNT here, it is probable that another panic will
+        // occur, sending us into an infinite panic loop that never calls the panic handler.
+        #[cfg(not(feature = "restricted-std"))]
+        if !must_abort {
+            panics = LOCAL_PANIC_COUNT.with(|c| {
```

Comment:
What does "likely" mean here? Your PR description says
> When restricted-std is in use, we do not have access to thread local storage, as sys/unsupported does not implement it. 

Is that *always* the case when using restricted-std? and if not, when can the behavior be different?

Reply:
Looking at it a bit more, it seems it's true _most_ of the time, but not necessarily always. [The restricted-std requirement is defined here if a platform match is not found](https://github.com/rust-lang/rust/blob/c97b539e408ea353f4fde2f9251d598291fec421/library/std/build.rs#L48), and since [thread_local is unimplemented in sys/unsupported](https://github.com/rust-lang/rust/blob/c97b539e408ea353f4fde2f9251d598291fec421/library/std/src/sys/unsupported/thread_local_key.rs#L5), a panic loop occurs.

There do appear to be some cases where this may not _always_ be true - i.e. in the case of `aarch64-apple-tvos`, which is [not allowed](https://github.com/rust-lang/rust/blob/c97b539e408ea353f4fde2f9251d598291fec421/library/std/build.rs#LL44-L44C13) to use regular unrestricted std, despite being [part of the `unix` family](https://github.com/rust-lang/rust/blob/c97b539e408ea353f4fde2f9251d598291fec421/compiler/rustc_target/src/spec/apple_base.rs#L150) and thus [using the `unix` system module](https://github.com/rust-lang/rust/blob/c97b539e408ea353f4fde2f9251d598291fec421/library/std/src/sys/mod.rs#L30) which does properly implement thread-local support. I would imagine this same situation would also apply to any of the other targets which are called out in the same area in `build.rs` _and_ also use a non-`unsupported` system module.

I'm not really sure how big of an impact this would be; since `restricted-std` is infectious, requiring itself to be in every std-dependent crate throughout the entire dependency tree in order to be used, it seems like a minor detail, especially since a double panic on these platforms will still trigger an abort via the global panic counter.